### PR TITLE
[BUGFIX #484] ensure transactions are always.

### PR DESCRIPTION
**Pull request by stefanpenner**
- [x] need to see if this actually fixes the problem
- [x] tests
- [x] review

### CODE REVIEW

**Comment by stefanpenner** [comment]
Although I think ^ should be added, this wont address all cases as.

`begin` -> `commit` are actually the public API for interacting with then VM, so code from the calling library between they two may still fail. Which means that could also needs the `try/finally`

**Comment by stefanpenner** [comment]
Rather then enforcing the caller to be aware of this. I change the PR to take a different approach.

* commit always resets the active transaction to null
* calling begin with an active transaction warns, then commits that transaction, and then starts the next transaction.

**Comment by stefanpenner** [comment]
There are different ways we can skin this cat, would love others opinions.

File: packages/@glimmer/runtime/lib/environment.ts [hunk 1]
```diff
@@ -330,7 +330,12 @@ export abstract class Environment {
   }
 
   begin() {
-    assert(!this._transaction, 'Cannot start a nested transaction');
+    let transaction = this._transaction;
+    if (transaction) {
+      this._transaction = null;
+      LOGGER.warn('a glimmer transaction was begun, but one already exists. You may have a nested transaction');
```
Comment (stefanpenner) [inline packages/@glimmer/runtime/lib/environment.ts]: throw in debug

**Comment by krisselden** [comment]
This looks good

**Comment by krisselden** [comment]
I like the fix, I'm not crazy about the test, why not throw on a component hook once like didInsertElement and test that you can rerender again?

**Pull request merged**